internal/engine/build_and_deployer.go

```diff
@@ -22,3 +23,46 @@ type BuildAndDeployer interface {
 	// consider putting elsewhere or renaming (`Warm`?).
 	GetContainerForBuild(ctx context.Context, build BuildResult) (k8s.ContainerID, error)
 }
+
+type FirstLineBuildAndDeployer BuildAndDeployer
+type FallbackBuildAndDeployer BuildAndDeployer
+
+type CompositeBuildAndDeployer struct {
+	containerBaD BuildAndDeployer
+	imageBaD     BuildAndDeployer
+}
+
+func NewCompositeBuildAndDeployer(firstLine FirstLineBuildAndDeployer, fallback FallbackBuildAndDeployer) *CompositeBuildAndDeployer {
+	return &CompositeBuildAndDeployer{
+		containerBaD: firstLine,
+		imageBaD:     fallback,
+	}
+}
+
+func (composite *CompositeBuildAndDeployer) BuildAndDeploy(ctx context.Context, service model.Service, currentState BuildState) (BuildResult, error) {
+	br, err := composite.containerBaD.BuildAndDeploy(ctx, service, currentState)
+	if err == nil {
+		return br, err
+	}
+	if shouldSkipImageBuild(err) {
+		return BuildResult{}, err
+	}
+	return composite.imageBaD.BuildAndDeploy(ctx, service, currentState)
+}
+
+// shouldSkipImageBuild determines whether the given error (from the containerBuildAndDeployer)
+// means that we should skip the image build.
+func shouldSkipImageBuild(err error) bool {
+	if _, ok := err.(*build.PathMappingErr); ok {
```

Comment:
it might be more idiomatic to have some sort of interface to indicate that this is an InvalidServiceError?